proposals/p2029.md

```diff
@@ -0,0 +1,103 @@
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+C++ reserves certain sets of identifiers for implementation-specific use. Carbon
+must ensure it respects these reserved identifiers to enable seamless and
+error-free interoperability with C++. Equally, it would be prudent to define a
+set of identifiers which are reserved for internal Carbon implementation usage.
+
+## Background
+
+The full set of C++ reserved identifiers is listed
+[here](https://en.cppreference.com/w/cpp/language/identifiers). This proposal is
+solely considering the set of user-definable identifiers which are considered
+names in C++:
+
+-   Identifiers with a double underscore anywhere are reserved;
+-   Identifiers that begin with an underscore followed by an uppercase letter
+    are reserved;
+-   Identifiers that begin with an underscore are reserved in the global
+    namespace.
+
+Considerations for reserved keywords are discussed in
+[p0093](https://github.com/carbon-language/carbon-lang/pull/93).
+
+## Proposal
+
+Carbon reserves the following identifiers in the specified contexts:
+
+-   Identifiers with a double underscore anywhere are reserved;
+-   Identifiers that begin with an underscore followed by an uppercase letter
+    are reserved.
+
+Carbon toolchains are expected to emit an error when a reserved identifier is
+encountered. This will remove the risk of undefined / unexpected runtime
+behaviour, or name resolution clashes when linking with C++ libraries.
+
+## Details
+
+This proposal makes use of three assumptions:
+
+1. A Carbon package exporting a C++ header will be scoped in a C++ namespace
+   following the package name;
+2. All C++ imports to Carbon will exist within the package `Cpp` which
+   represents the root of the C++ global namespace.
+3. The default Carbon package can not be exported to C++, following the
+   [no other library can import this library even from within the default package](/docs/design/README.md#files-libraries-packages)
+   rule.
+
+The stated reserved identifiers for Carbon ensure that there will be no
+user-defined identifier clashes between Carbon and C++ in either
+interoperability direction. As Carbon does not have a global namespace, and
+assumption 2 above explicitly gives the C++ global namespace a named scope
+within Carbon, there is no need to consider a Carbon equivalent of the C++
+_Identifiers that begin with an underscore are reserved in the global namespace_
+reservation.
+
+## Rationale
+
+This proposal supports the
+[Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+goal. It also provides a reserved identifier space for Carbon internal
+implementation specifics.
+
+## Alternatives considered
```

Comment:
My leaning would be to not do anything here. If desired, much as some Carbon features cannot be directly exposed to C++ ([requiring bridge code](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/interoperability/philosophy_and_goals.md#minimize-bridge-code)), code with C++-reserved identifiers could be treated similarly.

An advantage of that approach is that it avoids constraining Carbon's expressiveness based on what can be expressed in C++. While I think that may not be noticed short-term, long-term I think it could hinder Carbon's evolution.

@josh11b mentions C++ keywords above: I see those as a more significant evolution issue. It's worth noting that in the future C++ may add more keywords that Carbon doesn't. If the conclusion is that Carbon should reserve any reserved name in C++, it puts Carbon on the hook for preventing identifiers that match keywords not just in _current_ C++ standards, but _future_ C++ standards: a change to _C++_ could render Carbon code invalid, which I would lean away from.

Unlike C++, Carbon can also place identifiers that it may need in a package that it owns (the core library, or similar). It can also create anonymous identifiers. It shouldn't need to reserve identifiers in every package for this purpose.

A disadvantage of this approach is that, yes, exporting Carbon names that use C++ language-reserved identifiers will require a manual rename. However, my own leaning is that this is a reasonable disadvantage and that developers trying to write C++-interoperable code won't be particularly surprised by the consequence. This avoids the concerns of an automatic rewrite causing a conflict.

Reply:
OK, does this need to be a question for leads to work out the approach? On Discord there was interest in a short proposal to cover this, but happy to bin it if the collective view is otherwise.

On the reserved C++ keywords issues, the stalled PR #93 I link in the background section suggestions options such as warning programmers if they have used a reserved C++ keyword. Maybe this PR needs revisiting to nail this behaviour down?

Reply:
The only context I have is from [this discussion](https://discord.com/channels/655572317891461132/999035933959671808/1007120490286035037) -- if there's something more than that, I missed it. chandlerc did mention thinking a reserved space may be helpful, but I'm not sure what the intended use was -- if it was to match C++ or something else.

I don't know what the leads would prefer here. One option may be to just keep going with the proposal and see the response. i.e., it would be reasonable to mention my suggested alternative, offer up advantages and disadvantages, and see what the leads decide. A leads question may yield a similar result with possibly less effort, though they might still want to see the proposal.

Regarding #93, I think that's only mentioning possible uses, it's not trying to decide on them. Even if #93 had been accepted, I don't think it'd give you an answer. It's really a distinct question of whether Carbon leads want Carbon to reserve all C++-reserved identifiers. You could as easily add C++ keywords to the list of reserved identifiers here, which seems consistent with the current rationale of reserving things which would have name resolution clashes with C++.

But FYI both zygoloid and chandlerc probably won't respond this week.

Reply:
So I think it's a non-issue for lambdas (following an approach along the lines you suggested in the discussion), which was where I first raised the question, but it raised the wider discussion about having a reserved identifier space for C++ interop.

But I think your comments on this are intersecting with @josh11b's above - I'll add a section on keywords with some  options.